crates/p2p/src/swarm/mod.rs

```diff
@@ -571,44 +574,55 @@ impl P2P {
                 Ok(())
             }
             Command::ConnectToPeer(connect_to_peer_command) => {
-                // Whitelist peer
                 self.swarm
                     .behaviour_mut()
                     .allow_list
                     .allow_peer(connect_to_peer_command.peer_id);
 
-                // Add the peer to our config lists.
                 self.config.allowlist.push(connect_to_peer_command.peer_id);
-                self.config
-                    .connect_to
-                    .push(connect_to_peer_command.peer_addr.clone());
-
-                // Add peer to swarm
-                self.swarm.add_peer_address(
-                    connect_to_peer_command.peer_id,
-                    connect_to_peer_command.peer_addr.clone(),
-                );
+                for addr in &connect_to_peer_command.peer_addrs {
+                    self.config.connect_to.push(addr.clone());
+                    self.swarm
+                        .add_peer_address(connect_to_peer_command.peer_id, addr.clone());
+                }
+
+                let mut sorted_addrs = connect_to_peer_command.peer_addrs.clone();
+                sorted_addrs.sort_by_key(|a| {
+                    let s = a.to_string();
+                    if s.contains("/quic") { 0 } else { 1 }
+                });
 
                 let dialing_opts: DialOpts = DialOpts::peer_id(connect_to_peer_command.peer_id)
                     .condition(PeerCondition::DisconnectedAndNotDialing)
-                    .addresses(Vec::<Multiaddr>::from([connect_to_peer_command
-                        .peer_addr
-                        .clone()]))
+                    .addresses(sorted_addrs)
                     .extend_addresses_through_behaviour()
+                    // Only dial one address at a time for this peer to avoid parallel connection
+                    // attempts. This preserves the order of address dialing.
+                    .override_dial_concurrency_factor(NonZeroU8::new(1).unwrap())
```

Comment:
I think that if we do a very robust partition of the `Multiaddr` with `protocol_stack` as above we might not need this.

I'm not sure but worth a try.